src/main/c/truffleposix/truffleposix.c

```diff
@@ -73,6 +74,15 @@ struct truffleposix_stat {
 
 static void copy_stat(struct stat *stat, struct truffleposix_stat* buffer);
 
+char* truffleposix_get_user_home(const char *name) {
+  struct passwd *entry = getpwnam(name);
+  if (entry != NULL) {
+    return entry->pw_dir;
```

Comment:
This looks dangerous. The `pw_dir` value will change if the `getpwnam` is called again with a different `name` value. Maybe making a defensive copy or just using `getpwnam_r` would be better.

Reply:
Or is it that the `char*` is converted to a `java.lang.String` due to interop? In that case, we'd have our defensive copy anyway. But I'm also not sure we want to rely on that.

Reply:
No, we just treat `:string` as pointer arguments for NFI, and we read the String ourselves:
https://github.com/graalvm/truffleruby/blob/e06192899893238aa5c23597add65b015bdb9f3f/src/main/ruby/core/posix.rb#L107-L108
That makes a copy of course, we cannot have a `byte[]` of native memory.

Reply:
We could have a `byte[]` of native memory using `NativeRope`. I was under the impression we were using that more pervasively to reduce memcpy. But, if we're making a copy here, my core concern is covered.

Reply:
But you are right, `getpwnam` is not thread-safe (the JNR version was not either).
We'd need `getpwnam_r()` for that I guess.

Reply:
> We could have a `byte[]` of native memory using `NativeRope`.

NativeRope do not have a Java `byte[]`, but just a pointer to native memory.

Reply:
Right. But this function returns a `char*` and that was my point. If we were making a `NativeRope` we could end up with a pointer that changes underneath the rope.

Reply:
Let's address `getpwnam_r` in another PR, coming right up :)